service/epoch_lock.ml

```diff
@@ -14,10 +14,16 @@ type ('a, 'key) t = {
 }
 
 let activate t epoch ~ready ~set_ready =
-  t.current <- `Activating ready;
-  t.create epoch >|= fun v ->
-  t.current <- `Active (epoch, v);
-  Lwt.wakeup_later set_ready ()
+  Lwt.finalize
+    (fun () ->
+      t.current <- `Activating ready;
+      t.create epoch >|= fun v ->
+      t.current <- `Active (epoch, v);
+      Lwt.wakeup_later set_ready ())
+    (fun () ->
+      match t.current with
+      | `Activating _ -> Lwt.return (t.current <- `Idle)
+      | _ -> Lwt.return_unit)
```

Comment:
```ocaml=
if t.current = `Activating then t.current <- `Idle;
Lwt.return_unit
```

Reply:
This is the type ``| `Activating of unit Lwt.t``. 

I tried your suggestion and I got type issues.

Reply:
You're right, sorry. Forget about it.